### ISSUE TITLE
[BUG] gpg renderer failing to render some encrypted pillars

### ISSUE THREAD

**Issue by raddessi** [saltstack/salt#59307]
**Description**
I'm having issues where even though a set of gpg keys is setup up properly and salt can decrypt files on disk with the same encrypted contents as the pillars via running `gpg.decrypt` only _some_ of the pillars from the master are actually decrypted. I have seen no relevant log entries after setting the master to debug log level so I would appreciate some help in troubleshooting the gpg renderer.

This issue started after I upgraded to salt 3002, but reverting back to previous versions I was running just fine have not fixed it so that may just be coincidental. I had posted to the NTC slack [here](https://saltstackcommunity.slack.com/archives/C8832QN4U/p1610643802003100) but not much came of that.

**Setup**
`example_pillar1.sls`:
```
my_secret: | 
    -----BEGIN PGP MESSAGE-----

    <snipped>
    -----END PGP MESSAGE-----

```

`example_pillar2.sls`:
```
my_other_secret: | 
    -----BEGIN PGP MESSAGE-----

    <snipped>
    -----END PGP MESSAGE-----

```

**Steps to Reproduce the behavior**
Just run the master/minion as usual. Both of the pillar files are rendered, but only one is descrypted. Both are encrypted with the same key, and both of the encrypted contents WILL decrypt if placed on a file on disk and salt is ran with the `gpg.decrypt` module against that file.

**Expected behavior**
All secrets should be decrypted from the same key, right?

**Screenshots**
na

**Versions Report**
Running in a conda env, I have tried 3002.x, 3001.x, and 2019.2.x. I also upgraded from f31 to f32 with no change.
```
(salt) ~ # ❯❯❯ salt --versions-report
Salt Version:
          Salt: 3002.2
 
Dependency Versions:
          cffi: 1.14.4
      cherrypy: Not Installed
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.11.2
       libgit2: 1.1.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: 2.6.1
  pycryptodome: 3.9.9
        pygit2: 1.4.0
        Python: 3.8.5 (default, Sep  4 2020, 07:30:14)
  python-gnupg: 0.4.6
        PyYAML: 3.13
         PyZMQ: 20.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.3
 
System Versions:
          dist: fedora 32 
        locale: utf-8
       machine: x86_64
       release: 5.9.16-100.fc32.x86_64
        system: Linux
       version: Fedora 32 
 

```
</details>

**Additional context**
na

**Comment by network-shark** [comment]
Hello @raddessi 

I could only get it working with reading from files.
My workaround at the moment is.


```
#!jinja|yaml|gpg
 {% import_text "ssh_secret.gpg" as a_secret %}

ssh-secret3: {{ a_secret|yaml_encode }}
```

**Comment by raddessi** [comment]
Yeah.. this should be possible though, and was working just fine before. It has to be a bug somewhere, right now im suspecting within salt but im not sure how to troubleshoot the renderer.

**Comment by raddessi** [comment]
Ok.. via dogging through code I am pretty certain _most_ of my other pillar files are being processed corrcectly, but for some reason one or more pillar _files_ are just not at all being sent to the gpg renderer even though it is specified in the shebang. By adding logging to the gpg renderer I have looked for the data going in and coming out, but I am not seeing the file I'm having issues with even being passed in to the renderer. Where does salt handle processing each pillar file through the requested renderers?

**Comment by raddessi** [comment]
OK more progress. I can confirm gpg rendering does work, but the problem is that for some reason even though the gpg renderer is requested in the shebang of the pillar file, it is not ran. I have _temporarily_ patched this issue via appending `  - gpg: {}` to me list of ext_pillars in my master's config file. This is obviously not ideal though. I can assist in writing up a PR if you would point me to the section of code that may be responsible.. my searching has not helped me track that down yet.

The `ext_pillar` config on my master when this issue started was:
```
ext_pillar:
  - git:
    - master git@github.com:<redacted>
      - name: <redacted>
      - user: <redacted>
      - root: <redacted>
      - privkey: <redacted>
      - pubkey: <redacted>
  - netbox:
      api_url: <redacted>
      api_token: <redacted>
```

And now I have appended the gpg renderer to this to force the gpg renderer to run across each file:
```
ext_pillar:
  - git:
    - master git@github.com:<redacted>
      - name: <redacted>
      - user: <redacted>
      - root: <redacted>
      - privkey: <redacted>
      - pubkey: <redacted>
  - netbox:
      api_url: <redacted>
      api_token: <redacted>
  - gpg: {}
```

I'm not sure if it matters but all of the pillars I'm having issues with are in the git repo which is just my personal salt repo, it is set to pull the pillars from git since I run out of git directly and they are not easily present on the filesystem. I've been on the fence about changing back to filesystem backed since running out of git has had a number of issues but I haven't switched back yet.

**Comment by raddessi** [comment]
FR https://github.com/saltstack/salt/issues/58889 would have saved a bunch of my hosts from corrupting some of their config after writing out unencrypted data, FWIW.

**Comment by garethgreenaway** [comment]
@raddessi Thanks for the report.  So far I haven't been able to reproduce this one.  Can you provide a bit more detail about how things are setup.  Where are the pillar files being kept?  How are you including each pillar file?  Are they both in the top.sls file?  Or is one being included in another one?  Thanks!

**Comment by raddessi** [comment]
Yeah I think this one is going to be hard to reproduce, I'm not even sure what is triggering it yet.

The pillar filesystem structure is essentially:
top.sls:
```
base:
  '*':
    - example_pillar1
    - example_pillar2
```

And the example pillars are set up as shown above but with more data. Some files are sent to the gpg renderer and some are not, even though both of them have the same shebang. One of the pillar files that is not being sent to the gpg renderer is 42kB, not very large if you were worried about file size. All of the pillars are in the ext_pillar git repo, there are none on the local filesystem, I currently run entirely out of git.

Is there something specific you would like?

**Comment by garethgreenaway** [comment]
Thanks. That's good information to go on. When I tested i did not have the pillar files stored in Git.

**Comment by raddessi** [comment]
When I have time, I'll copy the git repo down locally and temporarily change to run out to a local dir and see if I can get it to fail the same way when using a local source rather than git, but I'm not sure if that will be this week or not.

**Comment by vrubiella** [comment]
Hello, I having similar issue only when run state with multiple targets (-L , wildcard,etc..) when state is raised only to one target GPG is always decrypted correctly.

I have state where GPG password is stored on .ini file and raise the state over a list of servers, some servers stores GPG password without decryption, and some servers store OK (no changes).
Consecutive execution with exactly same command  return different servers who fails decrypting GPG.

My workaround is run every server with only one target under for loop.... v_v. 
`for s in $(cat list|paste -sd',' -); do salt s state.sls  xxxx; done`

**Comment by garethgreenaway** [comment]
Circling back on this one.  Is the issue visible when doing a `pillar.items` or a `pillar.get` on a single item, or when attempting to use the pillar values in a state file?  Having two pillar files in Git, with GPG encrypted values in the pillar files, and doing a `pillar.items` so far I haven't seen the issue.

**Comment by raddessi** [comment]
It was consistently only rendering certain files with the gpg renderer so `pillar.get`, `pillar.item`, `pillar.items`, and any states that used those pillars would see the unencrypted data for the pillars that should have been decrypted but were not. It did seem at the time that the first few were being rendered just fine and the failing ones were always rendered "later" in the pillar top file.. but I can't really quantify that in any useful way at the moment. In order to add the major missing feature of pillar stacking to salt I recently moved all my pillars over to [salt-tower](https://github.com/jgraichen/salt-tower) (it's amazing :rocket:) and I have not checked if this is still needed yet with that specific configuration.. but in some cases salt-tower does use a slightly different rendering path than salt's defaults so I'm not sure if this issue would carry through. I will get back to this eventually but the next month will be fairly busy I think, I apologize for the delay in testing more on my end.

**Comment by raddessi** [comment]
> It was consistently only rendering certain files with the gpg renderer 

When I say it was only rendering some of the files with the gpg renderer  I don't mean only some files were being decrypted but all of them were passing through the renderer - only some of the pillar files that should have been logged as passing through the gpg renderer actually had the renderer run on them. The issue was not a failure of the gpg renderer to decrypt the data but rather salt was not sending the data through the gpg renderer at all in some cases which is why my temp fix of forcing the gpg renderer to run across _all_ my pillars works (in my case at least).

**Comment by raddessi** [comment]
Also, I apologize, part of my initial post was a little unclear. I never tested with a brand new empty master with only 2 pillar files, I did test with 2 almost identical pillar files with the contents I posted above but this was on my existing master with about 150 existing pillar files. I don't recall the exact pillar file names as I had placed them in the execution order closest to the pillars that were failing but I think one was nearer the start (the one that worked) and the other was near the end (the one that failed).

**Comment by garethgreenaway** [comment]
@raddessi All good.  Just trying to get to the scenario where the bug is surfacing 😃 So far I'm up to 20 pillar files with a single GPG encrypted pillar item in each file and haven't seen the issue manifest yet.  Do you have multiple encrypted values in the files?

**Comment by raddessi** [comment]
In probably more than half of them I do yes, but when I was testing this I tried to make it as simple as possible so for those two test pillars no it was just a single entry in each.

**Comment by raddessi** [comment]
I just remembered which pillar I noticed this failure in first, it was my users pillar with all my user config :) It's about 42k and at the time of this issue was using the shebang `#! yaml|gpg` and consisted of mostly this structure:

```yaml
users:
  username:
    top level options: blah
    also some deeper options:
      foo: baz
```
I fear that the issue was caused not by the pillars that failed to get sent to the gpg renderer but instead one of the previous pillars somehow.

**Comment by garethgreenaway** [comment]
@raddessi Is it possible that one of the pillar files failed and that caused the ones after that in the list to not be sent to the renderer?

**Comment by garethgreenaway** [comment]
So far I can reproduce this if the encrypted string is not in the right format or is missing some pieces, eg. the GPG decrypt has failed so it just shows the string.

**Comment by raddessi** [comment]
> @raddessi Is it possible that one of the pillar files failed and that caused the ones after that in the list to not be sent to the renderer?

It's possible a file failed in some way, for sure, but I never saw any evidence of it. Also.. I don't understand how a failure in one file could stop only one renderer from further execution.. but this is why I opened the issue. I'm just not familiar enough with the inner functionality of salt yet.

> So far I can reproduce this if the encrypted string is not in the right format or is missing some pieces, eg. the GPG decrypt has failed so it just shows the string.

This was not my case though, I tested with the same encrypted data in my test pillar files. and one was decrypted and the other not. Also, all of my data was valid and from the same encryption key. When the gpg renderer is forced to run across each file via my temp fix it decrypts all the encrypted items just fine so I have no good theory about what is actually causing the issue.

**Comment by esilva-everbridge** [comment]
FWIW we just started seeing this sort of behavior and there does not seem to be any consistency in which pillar values fail for any given minion. I can't help but think there might be an issue in gnupg2/gpg-agent, but have yet to be able to find anything concrete. The errors regarding "gpg: can't connect to the agent: IPC connect call failed" in the salt master logs are what lead me to think the gpg-agent is at issue.

**Comment by raddessi** [comment]
Yeah @esilva-everbridge that is the same experience I had. My workaround was posted in https://github.com/saltstack/salt/issues/59307#issuecomment-761709231 and is still working fine fwiw.. I have not remove that to test if it's still an issue yet I was waiting to hear back from this issue before testing again. What version of salt are you all running where you saw this issue? I think the last version I tested it on was 3003.1, we currently run 3004.

**Comment by raddessi** [comment]
We did also recently switch from centos-7 to alma-8.. I should test again on that.

**Comment by esilva-everbridge** [comment]
> 

We aren't using ext_pillar but I can test adding a reference to see if has an effect.

**Comment by ysagon** [comment]
Hi, I'm having the same issue. We aren't using ext_pillar as well as all the pillars we use are on disks.

I can confirm that adding

```
ext_pillar:
  - gpg: {}
```

to `/etc/salt/master` is a working workaround.

**Comment by raddessi** [comment]
@esilva-everbridge what OS did you get that on, and how did you enable verbose logging?

**Comment by esilva-everbridge** [comment]
I tried the `- gpg: {}` thing but that doesn't seem to make a difference.

@raddessi This is on Ubuntu 18.04. I did this to get logging going on the primary salt master (as root):

```shell
killall -9 gpg-agent
export GNUPGHOME=/etc/salt/gpgkeys
echo "verbose" >> $GNUPGHOME/gpg-agent.conf
echo "log-file $GNUPGHOME/agent.log" >> $GNUPGHOME/gpg-agent.conf
salt-call state.highstate
tail -f /etc/salt/gpgkeys/agent.log | grep -v terminated | grep -v started
```

**Comment by ysagon** [comment]
For me it's the only thing as well, but it is working. And not working without it.

**Comment by raddessi** [comment]
@esilva-everbridge do you happen to have `ext_pillar_first` set on your master?

**Comment by esilva-everbridge** [comment]
No. We do have these set:

```shell
pillar_merge_lists: True
pillar_source_merging_strategy: recurse
top_file_merging_strategy: merge_all
```

Our masters pull from different repositories, ones for the shared states/pillar and ones that are application specific. The shared states are the `base` saltenv, others have their own saltenv namespace. They are merged together.

This has worked well for us and we haven't seen any issues with it until recently and then only in a few environments. It isn't something I can easily recreate, which makes this really frustrating.

**Comment by raddessi** [comment]
Thanks for the context, ok good to know. That is really odd. This whole thing is, really.

Are all of the pillars that are failing to decrypt in the shared `base` saltenv by chance?

@garethgreenaway is there anything here that stands out to you?

**Comment by esilva-everbridge** [comment]
The failures seem to be in all saltenvs, but the initial highstates work fine, and not all highstates have failures.

**Comment by esilva-everbridge** [comment]
I don't know if it is relevant but the salt masters that are having issues are also using a FIPS kernel.

**Comment by raddessi** [comment]
I wonder if I'm just not seeing in time it but i think ours doesn't do that. Maybe i just do not have config files in the right spot

**Comment by ldav1s** [comment]
Have observed this on `salt-call` 3002.7 running on Ubuntu 20.04.3 LTS hosts.  Mostly decrypts the encrypted pillar alright, but will occasionally fail.

If it fails to decrypt the pillar it seems to swallow the error and just substitute the ciphertext instead of the decrypted value which is no good, IMO.  It should fail.

**Comment by raddessi** [comment]
@garethgreenaway since nobody can reproduce this yet on your end, could you all add a config kwarg to raise errors instead of failing silently if the pillar fails to decrypt? I would think the renderer would have to check for pgp signatures in the decrypted data and then raise an error if one is found after attempting decryption.

**Comment by ldav1s** [comment]
I was definitely in that group @raddessi .  Once I found the first incidence (which happened to write an entire small file) and checked the timestamps on that file across all the servers where that file is deployed, I realized it had been silently going on for a few servers at a time (at random) for _months_.

**Comment by raddessi** [comment]
Exactly. I had luckily found it in testing in my home lab before i saw it in prod.. 

In my mind the expected result from a gpg decryption failure should be an immediate halt. I don't want data trashed in prod, ever.

**Comment by msamadzadegan** [comment]
Hi,
I have problem with password encryption with those are in middle of string like:
 <Location />
                  AuthName ***
                  AuthType ***
                  AuthBasicProvider ***
                  AuthLDAPBindDN ***
                  AuthLDAPBindPassword ***
                  AuthLDAPURL ***
                  AuthLDAPGroupAttributeIsDN ***
 </Location> 

I use the the formula and pillar to create virtual host and when I put the encrypted password as value for "AuthLDAPBindPassword ", salt doesn't decrypt it
example:
 <Location />
                  AuthName ***
                  AuthType ***
                  AuthBasicProvider ***
                  AuthLDAPBindDN ***
                  AuthLDAPBindPassword -----BEGIN PGP MESSAGE-----
                                                         Version: GnuPG v2.0.22 (GNU/Linux)

                                                          hQEMAy1IbY+pMQYYAQf+MTv+40BwkJO3suA6RET7x2Z5OYTdpe4VveQpCHLNjFX+
                                                          +xF5nbbr2zKABQNMToErSvDiy6/02lwbueOlz+Ocnxp5HXHUwNpVIk/UZ9CX2U1e
                                                          cqHISza0A5NDqH7nFWRUShgB0fpIQZnAY+JNcJ2VMeGxWrTh1r+vQbERytGTT+qQ
                                                          R/+EhPHSrI1Lqe4Kf4JUUJ2m5/+qM6oxS9chRb/8aTIE4WEbnbnVGWNDlBnCRdg9
                                                          gdHRvfYxRyWe0MSEgcklRj01mQ6RcehTP1/2YtJh/W9SmtbDQZ2WnpIBvX262w08
                                                          2VinzcduAcpELmsUQU5b8SHC9OaXKoSYmnRgwvWco9JFAe5hEcPNs7IKdCgVmTq4
                                                          ngvg+FEkeo30rMhTp7zepLtzaPlYBnP6clIWU15KJdgpjdFabq0RnBWlyJKtpfJW
                                                          ZnqTSTLF
                                                          =gwTr
                                                          -----END PGP MESSAGE-----
                  AuthLDAPURL ***
                  AuthLDAPGroupAttributeIsDN ***
 </Location> 

When I test it in my test lab with the same version of master, minion and gpg agent that used in the production env it works but in production env it doesn't
is there any update?

**Comment by esilva-everbridge** [comment]
FWIW, I have only ever seen this under Ubuntu/Linux. My homelab is mostly FreeBSD and has never had this issue. I don't know if that helps or just makes things weirder, but there you go. This is the weirdest bug I have ever seen using Salt.

**Comment by dgengtek** [comment]
I had a similar issue with my gpg rendered files. The issue for me was that the pillar files were invalid and by using pygit2 any rendering issues are never shown in the logs which is why I was only able to find out about the issue by switching to pillar_roots. See the issue #59993

Try switching to pillar_roots temporarily and see if there are any errors logged on the salt master since you won't see any issues with git_pillar pygit2

**Comment by esilva-everbridge** [comment]
In my case we are not using gitfs and are using pillar_roots instead, so it seems like gnupg errors are just not logging at all.

I have also validated our secure pillar files to make sure there wasn't any issue in the formatting as well. We don't hand write our secure pillar files due to errors introduced using cut and paste from gnupg. I wrote a tool to manage the files: https://github.com/Everbridge/generate-secure-pillar
